src/k2/persistence/logStream/LogStream.cpp

```diff
@@ -0,0 +1,287 @@
+/*
+MIT License
+
+Copyright(c) 2020 Futurewei Cloud
+
+    Permission is hereby granted,
+    free of charge, to any person obtaining a copy of this software and associated documentation files(the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and / or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions :
+
+    The above copyright notice and this permission notice shall be included in all copies
+    or
+    substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS",
+    WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+    DAMAGES OR OTHER
+    LIABILITY,
+    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+*/
+
+#include "LogStream.h"
+#include <k2/common/Chrono.h>
+#include <k2/config/Config.h>
+#include <k2/dto/Collection.h>
+#include <k2/dto/Persistence.h>
+#include <k2/transport/RPCDispatcher.h>
+#include <k2/transport/RPCTypes.h>
+#include <k2/transport/Status.h>
+#include <k2/transport/TXEndpoint.h>
+#include <k2/dto/ControlPlaneOracle.h>
+#include <k2/dto/MessageVerbs.h>
+#include <k2/dto/LogStream.h>
+#include <cstdlib>
+#include <cctype>
+#include <algorithm>
+#include <random>
+
+namespace k2 {
+
+LogStream::LogStream() {
+    K2INFO("dtor");    
+}
+
+LogStream::~LogStream() {
+    K2INFO("~dtor");
+}
+
+seastar::future<> 
+LogStream::init(String cpo_url, String persistenceClusrerName){
+    K2INFO("LogStream Init");
+    _cpo = CPOClient(cpo_url);
+    return _client.init(persistenceClusrerName, cpo_url);
+}
+
+seastar::future<>
+LogStream::create(String logStreamName){
+    if (_logStreamCreate){
+        K2INFO("Log streaam already created");
+        throw std::runtime_error("Log stream already created");
+    }
+
+    _logStreamName = std::move(logStreamName);
+    _logStreamCreate = true;
+
+    // create metadata plogs in advance
+    std::vector<seastar::future<std::tuple<Status, String> > > createFutures;
+    for (uint32_t i = 0; i < METADATA_PLOG_POOL_SIZE; ++i){
+        createFutures.push_back(_client.create());
+    }
+    return seastar::when_all_succeed(createFutures.begin(), createFutures.end())
+    .then([this] (auto&& responses){
+        // register the metadata plog info to CPO
+        for (auto& response: responses){
+            auto& [status, plogId] = response;
+            if (!status.is2xxOK()){
+                throw std::runtime_error("unable to create plog for Metadata LogStream");

```

Comment:
We are trying to use our own exceptions, instead of generic runtime_error, so you can define a LogStreamException or something like that and throw it instead. Also, should there be some cleanup of plogs that are already created?

Reply:
added the own expection